[mod] typification of SearXNG: add new result type Code

return42 opened the pull request:
This patch adds a new result type: Code

- Python class:   searx/result_types/code.py
- Jinja template: searx/templates/simple/result_templates/code.html
- CSS (less)      client/simple/src/less/result_types/code.less

----

This PR is based on https://github.com/searxng/searxng/pull/5074

return42 committed:
[mod] typification of SearXNG: add new result type Code

This patch adds a new result type: Code

- Python class:   searx/result_types/code.py
- Jinja template: searx/templates/simple/result_templates/code.html
- CSS (less)      client/simple/src/less/result_types/code.less

Signed-of-by: Markus Heiser <markus.heiser@darmarIT.de>

Bnyro commented on searx/result_types/code.py:
```diff
@@ -0,0 +1,190 @@
+            # Wrap the code inside <pre> blocks using <code>, as recommended by
+            # the HTML5 specification (default is False).  Do we need this?
+            #
+            # kwargs["wrapcode"] = kwargs.get("wrapcode", True)
```
I think we should always set `wrapcode` to `True`.

It doesn't change the HTML formatting when looking at the result, but it semantically declares this paragraph as a `code` block. If we only used `pre`(formatted), it's not clear that this is actually source code.

And according to https://developer.mozilla.org/en-US/docs/Learn_web_development/Core/Accessibility/HTML
> Use semantic HTML, aka "The right element for the right job", because the browser provides so many built-in accessibility hooks.
>
Possibly, screen readers could also benefit from knowing its actual type?

return42 replied:
> I think we should always set `wrapcode` to `True`.

Thank you for keeping an eye on this; first I was unsure about my decision, but with you comment -> [done](https://github.com/searxng/searxng/pull/5131/files#diff-89850fe51ec4b76e79b5a89bb362f7080b339e5855ddf4a76be693f7267839a6R154)

Bnyro commented on searx/result_types/code.py:
```diff
@@ -0,0 +1,190 @@
+                # current code line
+                code_block = [code_line]
+                code_block_start = line_no
+                last_line_no = line_no
```
```suggestion
                previous_line_no = line_no
```
I've struggled a bit with understanding this code initially because I thought that `last_line_no` is meant to represent the last number of all codelines, e.g. (`6` if we have codelines `1,2,...,6`.

return42 replied:
Yes, the name is miss leading .. this is a pointer to the end of the *current* code block, I renamed the variable to: 

```
 code_block_end: int | None = None  # line where the current code ends
```

Bnyro commented on searx/result_types/code.py:
```diff
@@ -0,0 +1,190 @@
+                lexer = guess_lexer(src_code)
+            except ClassNotFound:
+                lexer = get_lexer_by_name("text")
+        return lexer
```
```suggestion
        src_code = "\n".join([l[1] for l in self.codelines])

        if self.filename:
            try:
                return guess_lexer_for_filename(self.filename, src_code)
            except ClassNotFound:
                pass

        try:
             return guess_lexer(src_code)
        except ClassNotFound:
             pass

        return get_lexer_by_name("text")
```
Perhaps this could be simplified by returning the found `lexer` early.

return42 replied:
Yes. makes sense / [done](https://github.com/searxng/searxng/compare/9c62d45688cb3253f8b7f8b88157e48314887483..77b5822990091f414076d3bb40336ff34fe7b94f)

Bnyro reviewed (commented):
Looks great all in all!

I've just been wondering if it wouldn't make more sense to display the file path / name directly above the formatted source code instead of the title, e.g. 
<img width="644" height="205" alt="Screenshot from 2025-08-30 11-47-18" src="https://github.com/user-attachments/assets/f20e1b98-3f79-4a7d-8284-cedcfcd3b051" />

Perhaps that's not related to this PR though?

return42 committed:
[mod] simple client: pygments.less - switched dark theme to monokai

Compared to ``lightbulb`` theme we used in the past for the dark theme, the
``monokai`` has a better contrast [1].

BTW, the result list of SearXNG should not act as a code-checker: The border &
color of class ``.err`` (used for syntax errors) has been removed / code
snippets are often not well formed nor valid code and the rendering of such code
fragments as errors obscures the view on the code snippet.

[1] https://pygments.org/styles/

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static
